### PR TITLE
feat(backend): Support authentication with ServiceAccountTokens. Part of #5138

### PR DESCRIPTION
**Pull request by elikatsis**
**Description of your changes:**
The changes are described in detail in this comment: https://github.com/kubeflow/pipelines/issues/5138#issuecomment-795391197

In a nutshell:
1. We make the authentication mechanisms extendable
2. We extend the authentication mechanisms with an authenticator for ServiceAccountTokens
3. The aforementioned authenticator expects ServiceAccountTokens
   * in a bearer token header: `Authorization: Bearer <token>`
   * with `ml-pipeline` as audience

/cc @Bobgy 
/cc @yanniszark 
/cc @StefanoFioravanzo 
/assign @elikatsis 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by Bobgy** [comment]
You also need to update `~/manifests/kustomize/base/installs/multi-user/api-service/cluster-role.yaml` to include API server role changes.

**Review by Bobgy — COMMENTED** [review]
Thank you for the super useful contribution that many users are hoping for!

I'm leaving some early comments before I finish reading all the code. I'll try to finish review tomorrow.
Note, most of them are nit pickings, so feel free to discuss about them. I don't think they are blockers.

File: backend/src/apiserver/server/experiment_server_test.go [hunk 1]
```diff
@@ -299,9 +306,12 @@ func TestListExperiment_Unauthorized(t *testing.T) {
 	clients, manager, _ := initWithExperiment_SubjectAccessReview_Unauthorized(t)
 	defer clients.Close()
 
+	userIdentity, err := manager.IsRequestAuthenticated(ctx)
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/server/experiment_server_test.go:309]: Actually we have been using `getUserIdentity()` in `getPermissionDeniedError()` (where we passed the context). These tests have to do with authorization, so they shouldn't raise any authentication issue, and we have other tests for the authentication part.

But your point makes sense. Some change may introduce errors and if we don't catch it with the dedicated tests we could hope for these to catch them :sweat_smile: 
I'll fix this.

File: backend/src/apiserver/auth/auth_util.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/metadata"
+)
+
+type Authenticator interface {
+	GetUserIdentity(ctx context.Context) (string, error)
+}
+
+// Make this public to reference it from tests
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/auth/auth_util.go:33]: Good idea, will do that

File: backend/src/apiserver/auth/auth_util.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+var IdentityHeaderMissingError = util.NewUnauthenticatedError(
+	errors.New("Request header error: there is no user identity header."),
+	"Request header error: there is no user identity header.",
+)
+
+// Make this public for tests to force its re-instantiation
+var Authenticators []Authenticator
+
+func GetAuthenticators(tokenReviewClient client.TokenReviewInterface) []Authenticator {
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/auth/auth_util.go:42]: Sounds good, I'll push a commit with that

File: backend/src/apiserver/common/const.go [hunk 4]
```diff
@@ -62,10 +62,14 @@ const (
...
 const (
-	GoogleIAPUserIdentityHeader string = "x-goog-authenticated-user-email"
-	GoogleIAPUserIdentityPrefix string = "accounts.google.com:"
+	GoogleIAPUserIdentityHeader    string = "x-goog-authenticated-user-email"
+	GoogleIAPUserIdentityPrefix    string = "accounts.google.com:"
+	AuthorizationBearerTokenHeader string = "Authorization"
+	AuthorizationBearerTokenPrefix string = "Bearer "
 )
 
+const TokenReviewAudience string = "ml-pipeline"
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/common/const.go:71]: We're good with this yep. I'll push a commit to change it.

Ideally the audience needs to be as unique as possible. This among with other potential security reasons is why it is common to use the name of the Kubernetes service backing the deployment.
However this may make more difficult the configuration of both the deployment of KFP and clients communicating to it.

Also, how about making it configurable via an environment variable (ConfigMap essentially) and have `pipelines.kubeflow.org` as a sane default?

File: backend/src/apiserver/server/util.go [hunk 5]
```diff
@@ -342,9 +311,9 @@ func isAuthorized(resourceManager *resource.ResourceManager, ctx context.Context
 	}
 
 	glog.Info("Getting user identity...")
-	userIdentity, err := getUserIdentity(ctx)
+	userIdentity, err := resourceManager.IsRequestAuthenticated(ctx)
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/server/util.go:314]: Yes definitely!

**Review by elikatsis — COMMENTED** [review]
@Bobgy thank you for your comments. They are to the point.
I've answered to each one of them and I'll push a few extra commits on top

File: backend/src/apiserver/common/const.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bobgy) [reply backend/src/apiserver/common/const.go:71]: Makes sense to me making it configurable with good default

**Comment by elikatsis** [comment]
@Bobgy I rebased on top of latest master (there were some conflicts after the merge of https://github.com/kubeflow/pipelines/pull/5231) adding extra commits to address your comments

File: backend/src/apiserver/server/util.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/server/util.go]: Ah, we need to have it public since it's called by `isAuthorized()` of package `server` ([source](https://github.com/arrikto/kubeflow-pipelines/blob/ac4936fbbfb44ff98a5267026925d7cfed1645e8/backend/src/apiserver/server/util.go#L314)) (same as `IsRequestAuthorized()`).
So it needs to be `AuthenticateRequest` instead of `authenticateRequest`.

File: backend/src/apiserver/auth/auth_util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bobgy) [reply backend/src/apiserver/auth/auth_util.go]: Sorry, I take my comment  back, this seems to be a common practice in go. However, when we do so, we should consider the error part of public interface of the module.

[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/auth/auth_util.go]: Sorry I'm a bit lost. So you mean to keep it as a public variable and move it to the `auth.go` file?

[the remark this answers is not in the document]
Comment (Bobgy) [reply backend/src/apiserver/auth/auth_util.go]: Yes, I meant it's OK to keep it as a public variable and move to `auth.go`. I investigated a bit about this approach and found it to be a fairly widely used pattern, so there's no need to be cautious of marking the public variables only usable in tests.

File: backend/src/apiserver/auth/auth.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+type Authenticator interface {
+	GetUserIdentity(ctx context.Context) (string, error)
+}
+
+// Make this public for tests to force its re-instantiation
+var Authenticators []Authenticator
+
+func GetAuthenticators(tokenReviewClient client.TokenReviewInterface) []Authenticator {
+	if Authenticators == nil {
+		Authenticators = []Authenticator{
```
Comment (Bobgy) [inline backend/src/apiserver/auth/auth.go:33]: I'm not very familiar with concurrent programming in golang. Can this be a race condition?
e.g. the server runs as many go routines, and there can be two incoming requests triggering this method at the same time.

/cc @capri-xiyue @elikatsis

File: backend/src/apiserver/auth/auth.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+			NewTokenReviewAuthenticator(
+				common.AuthorizationBearerTokenHeader,
+				common.AuthorizationBearerTokenPrefix,
+				[]string{common.GetTokenReviewAudience()},
+				tokenReviewClient,
+			),
+			NewHTTPHeaderAuthenticator(common.GetKubeflowUserIDHeader(), common.GetKubeflowUserIDPrefix()),
```
Comment (Bobgy) [inline backend/src/apiserver/auth/auth.go:40]: Shall we move HTTPHeaderAuthenticator above TokenReview? Because when we call them, we call them in order.
HTTPHeaderAuthenticator doesn't have any external dependency, so perf cost is minimal, when we call both.

File: backend/src/apiserver/auth/auth.go (hunk 6 above)
Comment (Bobgy) [reply backend/src/apiserver/auth/auth.go:33]: Do you have any concerns moving the init process to be part of server init?

**Review by Bobgy — COMMENTED** [review]
Thank you! I've went through the PR, it's very well structured and well written.
I feel pretty confident about it, except for some very minor nitpickings

**Comment by Bobgy** [comment]
Can you please resolve another conflict?

File: backend/src/apiserver/auth/auth_util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/auth/auth_util.go]: It feels a bit better to have it as a public variable, so I'll revert the change and move it to `auth.go` :)

File: backend/src/apiserver/auth/auth.go (hunk 6 above)
Comment (elikatsis) [reply backend/src/apiserver/auth/auth.go:40]: Hmm this could be an issue. Although I cannot think of a mesh of the threads' reads and writes that may result to bad values in `Authenticators` (the output is not related to the request, any request should result to the same authenticators being created), moving this to the initialization is probably a good idea.

But then we need to think what to do with this test (maybe drop it since testing the empty prefix shouldn't be part of the `auth_server` tests?): https://github.com/arrikto/kubeflow-pipelines/blob/3dc13dc773dcb240c8d15396b410051e221bfcb2/backend/src/apiserver/server/auth_server_test.go#L113

Comment (elikatsis) [reply backend/src/apiserver/auth/auth.go:40]: Oh I almost forgot. The main concern is what to do with the token review client. The reason we are passing it to this function is to be able to use a fake one for tests. Is it possible to we have a workaround for this?

File: backend/src/apiserver/auth/auth.go (hunk 7 above)
Comment (elikatsis) [reply backend/src/apiserver/auth/auth.go:47]: Sure!

File: backend/src/apiserver/auth/auth.go (hunk 6 above)
Comment (Bobgy) [reply backend/src/apiserver/auth/auth.go:40]: I think a simple way to avoid using a global variable as cache is to move `GetAuthenticators()` call to `NewResourceManager`: https://github.com/kubeflow/pipelines/blob/5990285894d655b0b52ca4d9a2f08756a99f1ec8/backend/src/apiserver/resource/resource_manager.go#L99

and store the authenticators as properties of resource manager.

In this way, they are only initiated once, also you can test them in tests using the fake.

Comment (Bobgy) [reply backend/src/apiserver/auth/auth.go:40]: the tricky thing about race condition is that, it's not only a matter of which order the two commands execute.
Both of them might be running at the same time and it can lead to memory corruption.

Although, I'm not sure whether memory corruption could happen when assigning an array to the same variable. In some tutorial the example was that when setting map key/values, two concurrent access can cause memory corruption.

Anyway, I'd prefer avoiding it, because it's not hard.

Comment (elikatsis) [reply backend/src/apiserver/auth/auth.go]: I pushed a commit with this change and tested it a bit. It looks like it works :tada: 
PTAL :)

**Comment by Bobgy** [comment]
/lgtm
Thanks! I feel this is ready, but I'd want to wait for @capri-xiyue's review too.
I think this can catch up with the release cut.

**Comment by Bobgy** [comment]
/approve
@elikatsis thank you for the great contribution!

**Comment by elikatsis** [comment]
@Bobgy and thank you for the quick and detailed review! I think the code is very improved now.
